docs/proposed/pub-sub.rst

```diff
@@ -0,0 +1,203 @@
+ - anyone wishing to subscribe is given a read-cap, "R"
+   - anyone possessing a valid read-cap "R" (including the publisher)
+     can obviously "invite" further clients
+   - nobody else may subscribe
+ - to add information, the publisher adds a new file to the mutable
+   directory
+ - old information can be removed (or not) as the publisher wishes
+
+Without further changes to Tahoe-LAFS, this can already be achieved:
+clients can "poll" the read-capability and determine if anything new
+exists.
+
+
+Improvements Over Polling
+-------------------------
+
+The above mechanism would scale somewhat poorly as the number of
+"broadcast" capabilities increased. It would be better if the
+storage-server could *tell* clients when a mutable Slot has changed.
+
+Storage servers implement a new feature, based on WebSockets. See also
+:ref:`http-storage-node-protocol` with which this protocol aims to be
+compatible. A single WebSocket endpoint exists:
+
+ - <server>/v1/mutable_updates
+
+After connecting to this endpoint, a client may send any number of
+messages (encoded using JSON) asking for updates to mutable
+Slots:
+
+    {
+        "tahoe-mutable-notification-version": 1,
+        "subscribe": [
+            "storage-index-0",
+            "storage-index-1",
+        }
+    }
+
+..where "storage-index-0" corresponds to an actual Storage Index of an
+existing mutable file. A client computes this from "R" (a read
+capability). The client must keep a mapping of `Storage Index ->
+read-capability` so it can match subsequent notifications.
+
+The server replies with a message like:
+
+    {
+        "tahoe-mutable-notification-version": 1,
+        "status": {
+            "storage-index-0": true,
+            "storage-index-1": "some kind of error message"
+        }
+    }
+
+That is, a "status" dict containing each requested Slot's
+storage-index with either "true" if the subscribe was successful or a
+string with an error-message if it was not possible to subscribe
+(e.g. the capability wasn't a mutable one, or wasn't found on this
+server). Every storage-index requested in the initial list will have
+an entry in this dict. The client should remember a map of all
+*successful* subscriptions back to the corresponding read-capability.
+
+TODO:
+
+ - for a given Slot, does a client just ask every storage-server for
+   the corresponding Storage Index? (are there information-disclosure
+   issues here?)
+
+       meejah: a storage-server can learn which Slots a client is
+       interested in, but it learns this information with the "polling"
+       mechanism too -- so i think the answer is, "to storage-server
+       learns the same information as with the polling method"
+
+ - do we want a more-structured response for errors, e.g. with
+   machine-readable "code" and a human-readable "message"?
+
+All subscriptions for a particular client shall be removed when that
+client's WebSocket connection goes away.
```

Comment:
Clients will have to contend with missed updates while they are disconnected.  This might not be a big deal for offline times that are planned/expected/intended and non-trivial in duration.  It might be a big deal for offline times due to transient network conditions or ongoing poor network quality.  If a client disconnects every couple minutes and has no way to receive updates for its time offline then it will basically have to fall back to polling.  Maybe these network conditions are worse than what we should be considering, I don't know.

This *could* be addressed by allowing for a small buffer on the storage-server and some kind of session identifier.  Reconnecting clients could try to re-establish their session and receive any missed updates and then only fall back to polling if they have been offline so long the storage server decided to forget about them.

If a server offered just a few minutes of buffer this would probably cover the majority of unintended disconnection.  The work saved in not having to service catch-up-polling presumably even offsets the cost of maintaining those buffers, up to some threshold.

It does add complexity to both the client and server implementation of pubsub and it doesn't even save any complexity on the client since clients must still fall back to polling if the server hasn't saved their session.  The performance benefits may be worth the extra complexity, though.

It also adds a possible DoS vector on storage servers since it allows clients to commit memory on servers and then release their own resources (disconnect and start a new garbage session, repeat).  This could be mitigated by accounting for the total number of sessions on the storage server when deciding how long a session lives, I suppose?  This changes the victim of the DoS from the storage server to any storage clients that are counting on their session to avoid having to poll.  And without a buffer they have to poll anyway, every time, so ... still a win?